Move from recording explicit positions to sortable integers

mootpointer opened the pull request:
This is a significant change (okay, it's pretty much a rewrite), but it make things more robust.

A lot of the techniques used here are inspired by [mixonic/ranked-model](https://github.com/mixonic/ranked-model).

Instead of using an explicit position stored against each row, we record an integer which allows the database to order things for us. That means that we have less database churn in many situations. Insertions will more frequently happen without the need for an update of other rows.

This is not yet ready to be merged, because the documentation in README.md and the module and the inline documentation has not been updated. I'd also be keen for some code review before I merge this.

mootpointer committed:
Move from explicit positioning, to ranking

mootpointer committed:
Handle scoped ranking

mootpointer committed:
Add handling for collisions

mootpointer committed:
Refactor a bit to placate credo

gogogarrett commented on lib/ecto_ordered.ex:
```diff
@@ -65,190 +58,270 @@ defmodule EctoOrdered do
+  defp neighbours_at_position(%Order{
+                                     rank_field: rank_field,
+                                     repo: repo
+                                    } = order, position, cs) when position <= 0 do
```
I am curious to see what the style guide for this type of stuff end up as. I see that [phoenix do this style](https://github.com/phoenixframework/phoenix/blob/master/lib%2Fphoenix%2Fchannel%2Fserver.ex#L72-L73) where they nest the guards on a new line

mootpointer replied:
Yeah. Credo doesn't bother me about it, but it would bother me about super-long lines.

In some of these places, I don't need to match in the function definition, but I was just following on the style from the existing library.

joshprice replied:
That Phoenix style is correct as confirmed by José himself

joshsmith commented on lib/ecto_ordered.ex:
```diff
@@ -34,23 +34,16 @@ defmodule EctoOrdered do
   """
 
+  @max 8388607
+  @min -8388607
```
Can you document this as being a signed MEDIUMINT? I wasn't sure what I was looking at here.

Also, is there a reason for that choice in particular?

mootpointer replied:
It was a hold-over from https://github.com/mixonic/ranked-model, but I'm thinking a BIGINT is going to be a better choice in the long term.

joshsmith commented on test/ecto_ordered_test.exs:
```diff
@@ -337,16 +328,11 @@ defmodule EctoOrderedTest.Scoped do
-    assert Repo.get(Model, xmodel3.id).scoped_position == 3
-    assert Repo.get(Model, model2.id).scoped_position == 4
-    assert Repo.get(Model, model2.id).scope == 2
+    assert ranked_ids(Model, 2) == [xmodel1.id, xmodel2.id, xmodel3.id, model2.id]
```
Not sure what the `xmodel` naming scheme means.

mootpointer replied:
Fair point. I'll re-name these to make this more clear. (It's a while since I authored this, but I'm guessing it was a holdover from the existing tests which I haven't got around to refactoring)

JoshSmith commented:
Looking really good here @mootpointer. Excited to see this and #4 get merged in! Just in time for some things we're working on, too.

mootpointer committed:
Fix issues arising with ecto2

mootpointer committed:
Increase the max and min values to use INT values

mootpointer committed:
Rename variables in scope test for clarity

joshsmith commented on test/ecto_ordered_test.exs:
```diff
@@ -317,21 +317,21 @@ defmodule EctoOrderedTest.Scoped do
-    xmodel3 = Model.changeset(%Model{scope: 2, title: "item #3"}, %{}) |> Repo.insert!
+    scope2_model1 = Model.changeset(%Model{scope: 2, title: "item #1"}, %{}) |> Repo.insert!
+    scope2_model2 = Model.changeset(%Model{scope: 2, title: "item #2"}, %{}) |> Repo.insert!
+    scope2_model3 = Model.changeset(%Model{scope: 2, title: "item #3"}, %{}) |> Repo.insert!
```
👍 way better!

mootpointer committed:
Add the ability to move rows up or down

This does not require you to know the rank of a row, but you can simply
specify a virtual `:move` field.

JoshSmith commented:
That last commit is looking great @mootpointer. Excited to see some of this.

I haven't noticed for sure – still too new to the project – but do you have or are you planning to have an interface for moving to top or bottom of a list?

mootpointer commented:
Yeah `move: :first` and `move: :last` are totally doable.

I'm thinking I might refactor all this first though, since working on these features wasn't ultra-fun due to some poor earlier design choices.

mootpointer committed:
Rename Order struct to Options for clarity

JoshSmith commented:
Yeah, totally go for refactoring, and feel free to handle in a separate PR, obviously. This one is already quite large.

I can open a separate issue if you'd like to track?

mootpointer commented:
Yeah. Crack open another issue. This can be an enhancement.

mootpointer committed:
Consolidate some of the query functions

mootpointer committed:
Clarify naming of shift_others_* functions

mootpointer committed:
Update docstrings

mootpointer committed:
Add default fields

mootpointer committed:
Ensure unranked items are ranked on update

mootpointer commented:
Okay. This is getting close to being ready to land. There's a couple of edges I need to sort out as well as docs – so many docs.

I've currently got this branch running in staging with real users so we'll find out how that goes.

zachdaniel commented:
Hey guys!  We're very interested in using this library, but we're on a bit of a time crunch.  Any way that we can help get this pr move along?

mootpointer commented:
Hey @zachdaniel, I just wanted to do some testing to make sure that when I release this into the wild no one gets hurt...

I've had it in production for a little bit and have found a couple of bugs. I could consider merging this and putting it in a beta or rc release, but I'm not yet happy putting it out as a proper release.

mootpointer committed:
Fix issue with rebalancing :last positioned items

JoshSmith commented:
@mootpointer could we put it into a beta and post some of the issues here? I'm sure between @zachdaniel and myself we might be able to help get it into a release-able state.

zachdaniel commented:
@JoshSmith :100:

mootpointer commented:
Okay. I'm going to merge this into master and release `0.2.0-beta1` which we can work towards releasing as `0.2.0` next week. How does that sound?

JoshSmith commented:
@mootpointer sounds great!

zachdaniel commented:
Sounds fantastic!

On Mon, Nov 7, 2016 at 9:19 PM Josh Smith <notifications@github.com> wrote:

> @mootpointer <https://github.com/mootpointer> sounds great!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zovafit/ecto-ordered/pull/5#issuecomment-259034492>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFdQ467PKd2cT41jnvhhoJ5UbPgflr9Zks5q7-odgaJpZM4IbzAJ>
> .
>
